bigtable-hbase-1.x-parent/bigtable-hbase-1.x-mapreduce/src/main/java/org/apache/hadoop/hbase/mapreduce/BigtableSyncTableAdapter.java

```diff
@@ -0,0 +1,384 @@
+      printUsage("--sourcetablename is required.", args);
+      return false;
+    }
+
+    if (syncTable.targetTableName == null) {
+      printUsage("--targettablename is required.", args);
+      return false;
+    }
+
+    return true;
+  }
+
+  /**
+   * create source connection with Bigtable
+   *
+   * @param mapper
+   * @param conf
+   * @throws IOException
+   */
+  public static void createBigtableSourceConnection(SyncMapper mapper, Configuration conf)
+      throws IOException {
+    // create source connection config
+    // inherit base config, but override connection based on job args
+    Configuration bigtableConf = new Configuration(conf);
+    bigtableConf.unset(ClusterConnection.HBASE_CLIENT_CONNECTION_IMPL);
+    BigtableConfiguration.configure(
+        bigtableConf,
+        bigtableConf.get(SOURCE_BT_PROJECTID_CONF_KEY),
+        bigtableConf.get(SOURCE_BT_INSTANCE_CONF_KEY),
+        bigtableConf.get(SOURCE_BT_APP_PROFILE_CONF_KEY, ""));
+
+    LOG.info(
+        "configuring source connection to Bigtable with configurations: "
+            + bigtableConf.get(SOURCE_BT_PROJECTID_CONF_KEY)
+            + ", "
+            + bigtableConf.get(SOURCE_BT_INSTANCE_CONF_KEY)
+            + ", "
+            + bigtableConf.get(SOURCE_BT_APP_PROFILE_CONF_KEY, ""));
+
+    try {
+      if (mapper.sourceConnection != null) {
+        mapper.sourceConnection.close();
+      }
+    } catch (IOException ioe) {
+      LOG.warn("closed source connection, " + ioe.getMessage());
+      ioe.printStackTrace();
+    }
+
+    mapper.sourceConnection = openConnection(bigtableConf);
+    mapper.sourceTable =
+        openTable(mapper.sourceConnection, bigtableConf, SyncTable.SOURCE_TABLE_CONF_KEY);
+  }
+
+  /**
+   * create target connection with Bigtable
+   *
+   * @param mapper
+   * @param conf
+   * @throws IOException
+   */
+  public static void createBigtableTargetConnection(SyncMapper mapper, Configuration conf)
+      throws IOException {
+    // create target connection config
+    // inherit base config, but override connection based on job args
+    Configuration bigtableConf = new Configuration(conf);
+    bigtableConf.unset(ClusterConnection.HBASE_CLIENT_CONNECTION_IMPL);
+    BigtableConfiguration.configure(
+        bigtableConf,
+        bigtableConf.get(TARGET_BT_PROJECTID_CONF_KEY),
+        bigtableConf.get(TARGET_BT_INSTANCE_CONF_KEY),
+        bigtableConf.get(TARGET_BT_APP_PROFILE_CONF_KEY, ""));
+
+    LOG.info(
+        "configuring target connection to Bigtable with configurations: "
+            + bigtableConf.get(TARGET_BT_PROJECTID_CONF_KEY)
+            + ", "
+            + bigtableConf.get(TARGET_BT_INSTANCE_CONF_KEY)
+            + ", "
+            + bigtableConf.get(TARGET_BT_APP_PROFILE_CONF_KEY, ""));
+    try {
+      if (mapper.targetConnection != null) {
+        mapper.targetConnection.close();
+      }
+    } catch (IOException ioe) {
+      LOG.warn("closed target connection, " + ioe.getMessage());
```

Comment:
Failed to close target connection? This is very confusing, because you are actually establishing "real" target connection later.

Reply:
updated this to debug level logging and updated msg, ptal